Docs/pluggable components (#2851)
dapr/docs: daprdocs/content/en/operations/extending/developing-grpc-components.md

---
type: docs
title: "Developing gRPC-based Pluggable Components"
linkTitle: "Developing gRPC-based Pluggable Components"
weight: 250

> description: "Extending dapr with external gRPC-based components"

Reviewer: ```suggestion
description: "Extending Dapr with external gRPC-based components"
```

---

> [gRPC-based](https://grpc.io/) Dapr components are typically run as containers or processes that communicate with the dapr main process via [Unix Domain Sockets](https://en.wikipedia.org/wiki/Unix_domain_socket).

Reviewer: It would be better to create context to state what gRPC Dapr components _are_ instead of saying _how_ they are typically used. As this is the first document in this series, this document should provide context of what is to follow if the reader choose to embark in this reading journey - or link to some place that would provide such.

Suggestion: Use reference links to avoid repeating the main GRPC link. Not really a big deal as you only link it twice but it would allow you to easily turn any textual reference to gRPC into a links.

Author: Agree, will work on that

## Developing a Pluggable Component from scratch

{{< tabs ".NET" >}}
{{% codetab %}}

### Step 1: Prerequisites

> For this tutorial, you must have minimal knowledge of [gRPC and protocol buffers](https://grpc.io/), and that you chose a programming language [that supports gRPC](https://grpc.io/docs/languages/).

Reviewer: ```suggestion
For this tutorial, you must:
1. Have some knowledge of how to use and build services using [gRPC and protocol buffers](https://grpc.io/). In doubt, check the their [Quick Start](https://grpc.io/docs/languages/csharp/quickstart/) and [Basics tutorial](https://grpc.io/docs/languages/csharp/quickstart/) .
2. Use a [gRPC-supported programming language](https://grpc.io/docs/languages/).
```

For simplicity, all code samples uses the generated code from the [protoc](https://developers.google.com/protocol-buffers/docs/downloads) tool or supported gRPC build tools by languages.

> As previously mentioned, dapr uses a Unix Domain Socket to communicate with gRPC-based components, which means that as a prerequisite you also need a UNIX-like system, or for Windows users, [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) should be sufficient.

Reviewer: I think we could move this particular paragraph to a language-agnostic section and provide links for .Net specific stuff later. What you think?

Author: Not sure, [I've found other documentation doing the same](https://github.com/dapr/docs/blob/v1.8/daprdocs/content/en/getting-started/quickstarts/serviceinvocation-quickstart.md), and I think it worth to pay for duplicated text against having two prerequisite sections?

> Also, install [.NET Core 6+](https://dotnet.microsoft.com/en-us/download)

Reviewer: ```suggestion
Install [.NET Core 6+](https://dotnet.microsoft.com/en-us/download)
```

> This tutorial is based on the [official Microsoft documentation](https://learn.microsoft.com/en-us/aspnet/core/grpc/basics?view=aspnetcore-6.0#c-tooling-support-for-proto-files).

Reviewer: ```suggestion
This tutorial is based on the [official Microsoft documentation for development using Protocol Buffers](https://learn.microsoft.com/en-us/aspnet/core/grpc/basics?view=aspnetcore-6.0#c-tooling-support-for-proto-files).
```

### Step 2: Downloading the base template project

> We've prepared a blueprint project that helps you to skip a few manual steps like cloning dapr repository and adding the required libraries. Get started by downloading and unzipping it using the following <a href="/code/dotnet-memstore.zip">link</a>.

Reviewer: ```suggestion
We've prepared a blueprint project that helps you to skip a few manual steps like cloning Dapr repository and adding the required libraries. Get started by downloading and unzipping it using the following <a href="/code/dotnet-memstore.zip">link</a>.
```

> This template contains all the required changes to start developing a gRPC-based component from scratch, including protobuf definitions and an unimplemented InMemory StateStore that you are going to implement.

Reviewer: ```suggestion
This template contains all the required changes to start developing a gRPC-based component from scratch, including protobuf definitions and an unimplemented in-memory StateStore that you are going to implement.
```

### Step 3: Building and running

At this point, you are ready to build and run the component. Let's test it out by running `dotnet build`

```

> ➜  DaprMemStoreComponent dotnet build

Reviewer: Nit: Not sure on windows-world but shell prompts are usually referenced by `$` instead of `➜`.

Author: yup, its my zsh prompt theme

MSBuild version 17.3.0+92e077650 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  DaprMemStoreComponent -> [redacted]

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.03
```

> Great!
> 
> Now, let's run the StateStore service by issuing a `dotnet run`

Reviewer: ```suggestion
Great! You've just built your first gRPC-based pluggable component: an in-memory StateStore component!

Now, let's run this StateStore service by issuing a `dotnet run`
```

```
➜  DaprMemStoreComponent dotnet run
Building...
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://localhost:5259, https://localhost:7089'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://unix:/tmp/dapr-components-sockets/memstore.sock
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
```

To see it working, you can use the [grpc_cli](https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) tool for making gRPC calls. You can download and use it or use your preferred tool.

> Once properly downloaded and installed, open a new terminal instance, and let's invoke the `Features` method from the StateStore service, using the following command: `grpc_cli call unix:///tmp/dapr-components-sockets/memstore.sock Features ''`

Reviewer: Out of curiosity: who is setting the DAPR_... env var? I am curious how the component got the `/tmp/dapr-coomponents-socket/memstore.sock` path. It wasn't given on the previous run step.

Author: from template, which actually contains a ready-to-use unimplemented state store including socket handling like creating/deleting

<!-- TODO why of gRPC pluggable components? -->
<!-- What is cool interesting about developing components as gRPC pluggable components? What are the benefits? -->

[gRPC-based][grpc] Dapr components are typically run as containers or processes that communicate with the Dapr main process via [Unix Domain Sockets][uds].

<!-- TODO how of gRPC pluggable components? -->

## Developing a Pluggable Component from scratch

{{< tabs ".NET" >}}
{{% codetab %}}

### Step 1: Prerequisites

For this tutorial, you must:

1. Have some knowledge of how to use and build services using [gRPC and protocol buffers][grpc]. In doubt, check the their [Quick Start](https://grpc.io/docs/languages/csharp/quickstart/) and [Basics tutorial](https://grpc.io/docs/languages/csharp/quickstart/).
2. Use a [gRPC-supported programming language](https://grpc.io/docs/languages/).
3. Have an operating system that supports Unix Domain Sockets.

For simplicity, all code samples uses the generated code from the [protoc](https://developers.google.com/protocol-buffers/docs/downloads) tool or supported gRPC build tools by languages.

> As previously mentioned, Dapr uses a Unix Domain Socket to communicate with gRPC-based components, which means that as a prerequisite you also need a UNIX-like system, or for Windows users, [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) should be sufficient.

Reviewer 2: ```suggestion
As previously mentioned, Dapr uses a Unix Domain Socket to communicate with gRPC-based components, which means that as a prerequisite you also need a UNIX-like system, or for Windows users, [WSL](https://learn.microsoft.com/windows/wsl/install) should be sufficient.
```

to fix the link validation error

Install [.NET Core 6+](https://dotnet.microsoft.com/en-us/download)

> This tutorial is based on the [official Microsoft documentation for development using Protocol Buffers](https://learn.microsoft.com/en-us/aspnet/core/grpc/basics?view=aspnetcore-6.0#c-tooling-support-for-proto-files).

Reviewer 2: ```suggestion
This tutorial is based on the [official Microsoft documentation for development using Protocol Buffers](https://learn.microsoft.com/aspnet/core/grpc/basics?view=aspnetcore-6.0#c-tooling-support-for-proto-files).
```

to fix the link validation error
